Reconcile split brain

Vagabond opened the pull request.

Vagabond committed:
Make riak_repl_leader:set_candidates a cast, so it doesn't block waiting for election

bz1146 zd773

Vagabond committed:
Trigger a leader election if there's 2 leaders after a netsplit

bz1146 zd773

Vagabond committed:
Speed up elections during a netsplit by skipping delayed candidate_timer events

bz1146 zd773

In a netsplit scenario, sending heartbeats to downed nodes can be very
expensive (have to wait for TCP timeouts). This causes candidate_worker
messages to stack up in the mailbox. This patch flushes them to prevent
that from happening.

jonmeredith commented:
I set up 4 VMs the one that was elected as leader I disconnected the virtual NIC and waited until I saw disconnected messages on all the nodes then re-enabled the NIC.   Once the nodes reconnected there were two leaders, i.e. no re-election took place.  This sounds very similar to your physical plugging test - it isn't a 'fiddle with the cookie' kind of a test.

Vagabond committed:
Have new leaders send checklead msgs; Don't omit up nodes from checkleads

bz1146 zd773

This resolves an issue where checklead messages weren't actually being
sent because erlang saw the node as up but gen_leader didn't.
Additionally, the leader elected on the other side of the split wasn't
sending checklead messages, this has also been resolved.

Vagabond committed:
Break a deadlock when starting a candidate

bz1146 zd773

When a candidate starts but the leader is behind it in the candidate
list but all nodes before it are up, it would sit in safe_loop forever.
This patch implements a fix to the issue.

Vagabond committed:
Remove io:formats

Vagabond committed:
Fix some indenting
